src/library/blast/blast.cpp

```diff
@@ -551,6 +580,62 @@ class blastenv {
         return m_abstract_expr_manager.hash(e);
     }
 
+    void init_imp_extension_entries() {
+        for (auto & p : get_imp_extension_manager().get_entries()) {
+            branch_extension & b_ext = curr_state().get_extension(p.second);
+            b_ext.inc_ref();
+            m_imp_extension_entries.emplace_back(p.first(), p.second, static_cast<imp_extension*>(&b_ext));
+        }
+    }
+
+    void finalize_imp_extension_entries() {
+        for (auto & e : m_imp_extension_entries) {
+            e.m_ext_of_ext_state->dec_ref();
+        }
+    }
+
+    list<imp_extension*> get_ext_path(imp_extension * _imp_ext) {
+        list<imp_extension*> path;
+        imp_extension * imp_ext = _imp_ext;
+        while (imp_ext != nullptr) {
+            path = cons(imp_ext, path);
+            imp_ext = imp_ext->m_parent;
+        }
+        return path;
+    }
+
+    imp_extension_state & get_imp_extension_state(unsigned state_id) {
+        lean_assert(state_id < m_imp_extension_entries.size());
+        imp_extension_entry & e = m_imp_extension_entries[state_id];
+        imp_extension_state * ext_state = e.m_ext_state.get();
+        imp_extension * ext_of_curr_state = static_cast<imp_extension*>(&curr_state().get_extension(e.m_ext_id));
+        lean_assert(e.m_ext_of_ext_state);
+        imp_extension * ext_of_ext_state = e.m_ext_of_ext_state;
+
+        list<imp_extension*> curr_state_path = get_ext_path(ext_of_curr_state);
+        list<imp_extension*> ext_state_path = get_ext_path(ext_of_ext_state);
+
```

Comment:
You are allocating a list for each `get_ext_path`, then you reverse one of them.
You can avoid this overhead by storing the path into a buffer<imp_extension*>.
Buffers can be easily traversed forwards/backwards with no overhead.
Our `buffer` class also avoids heap allocation if the number of elements is small.